### ISSUE TITLE
Conflict with @types

### ISSUE THREAD

**Issue by pocesar** [NativeScript/nativescript-dev-typescript#19]
When installing typings using `@types/packageName` and the package requires `@types/node`, it conflicts with declarations from `tns-core-modules`:

```
node_modules/@types/node/index.d.ts(38,11): error TS2300: Duplicate identifier 'MapConstructor'.

node_modules/@types/node/index.d.ts(40,11): error TS2300: Duplicate identifier 'SetConstructor'.

node_modules/@types/node/index.d.ts(49,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'global' must be of type 'any', but here has type 'Global'.
node_modules/@types/node/index.d.ts(73,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'require' must be of type 'NativeScriptRequire', but here has type 'NodeRequire'.
node_modules/@types/node/index.d.ts(85,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'module' must be of type 'NativeScriptModule', but here has type 'NodeModule'.
node_modules/@types/node/index.d.ts(3815,5): error TS2309: An export assignment cannot be used in a module with other exported elements.
node_modules/tns-core-modules/es-collections.d.ts(30,14): error TS2300: Duplicate identifier 'MapConstructor'.
node_modules/tns-core-modules/es-collections.d.ts(31,14): error TS2300: Duplicate identifier 'SetConstructor'.
```

is there any reason why `tns-core-modules` are using it's own version of the `node` typings?

**Comment by ivanpadavan** [comment]
I'm getting this the exact same error while trying to compile just generated angular2 template for android

**Comment by jarrodchesney** [comment]
I'm having this issue with a merged angular-cli project that uses protractor. I've had to take protractor out.

**Comment by Deon588** [comment]
i'm having this issue after uninstalling/re-installing angular-cli...  could it have something to do with clearing the npm cache...???

**Comment by bnussey** [comment]
Any updates on this?

**Comment by jarrodchesney** [comment]
I added these links to my tsconfig.json, this seemes to make the compiler not load any types, including the "@types/node"

I think if you needed some types, you could add these in.

https://github.com/Synerty/peek-client-fe/blob/master/peek_client_fe/tsconfig.json#L13
```
    "typeRoots": [
      "./node_modules/@types"
    ],
    "types": []
```

**Comment by ivanpadavan** [comment]
My project was in folder which was in folder with other project (native script inside angular-cli) . Problem dissapeared after replacing project in some other place. This issue could be overcomed also after reviewing path env var. Look carefully, this is about attention

**Comment by Deon588** [comment]
I resolved this by removing typescript and re-installing a lower version of Typescript...  It seems anything > v2.0.2 gives me this error...  strange...

**Comment by Deon588** [comment]
For me it worked after changing Typescript to V2.0.2  Anything higher than
that and the issue is there for some reason...

**Comment by hdeshev** [comment]
The real fix requires a lot of changes to the published NativeScript typings, some of which are not backwards compatible. It will require changes to client `tsconfig.json` files too -- we'll require `"lib": ["es6", "dom"]` as well. The latter should be handled automatically by an updated version of the nativescript-dev-typescript package.

All that will be available with the NativeScript 3.0 release. Until then, I'd advise that people use TypeScript 2.0.x or avoid referencing DOM or Node typings.

**Comment by KeithGillette** [comment]
I look forward to the permanent fix with NativeScript 3.0, as I am experiencing similar errors. FWIW, I have worked around them in NativeScript 2.5.1 by updating `tsconfig.sjon` with `"lib": ["es6", "dom"]` and modifying `references.d.ts` to include most but not all of `node_modules/tns-core-modules/tns-core-modules.d.ts`, as follows:

**references.d.ts**
```
/// <reference path="node_modules/tns-core-modules/application-settings/application-settings.d.ts" />
/// <reference path="node_modules/tns-core-modules/application/application.d.ts" />
/// <reference path="node_modules/tns-core-modules/camera/camera.d.ts" />
/// <reference path="node_modules/tns-core-modules/color/color.d.ts" />
/// <reference path="node_modules/tns-core-modules/color/known-colors.d.ts" />
/// <reference path="node_modules/tns-core-modules/connectivity/connectivity.d.ts" />
/// <reference path="node_modules/tns-core-modules/console/console.d.ts" />
/// <reference path="node_modules/tns-core-modules/css-value/reworkcss-value.d.ts" />
/// <reference path="node_modules/tns-core-modules/css/reworkcss.d.ts" />
/// <reference path="node_modules/tns-core-modules/data/observable-array/observable-array.d.ts" />
/// <reference path="node_modules/tns-core-modules/data/observable/observable.d.ts" />
/// <reference path="node_modules/tns-core-modules/data/virtual-array/virtual-array.d.ts" />
/// <reference path="node_modules/tns-core-modules/file-system/file-name-resolver.d.ts" />
/// <reference path="node_modules/tns-core-modules/file-system/file-system-access.d.ts" />
/// <reference path="node_modules/tns-core-modules/file-system/file-system.d.ts" />
/// <reference path="node_modules/tns-core-modules/fps-meter/fps-meter.d.ts" />
/// <reference path="node_modules/tns-core-modules/fps-meter/fps-native.d.ts" />
/// <reference path="node_modules/tns-core-modules/http/http.d.ts" />
/// <reference path="node_modules/tns-core-modules/image-asset/image-asset.d.ts" />
/// <reference path="node_modules/tns-core-modules/image-source/image-source.d.ts" />
/// <reference path="node_modules/tns-core-modules/js-libs/esprima/esprima.d.ts" />
/// <reference path="node_modules/tns-core-modules/location/location.d.ts" />
/// <reference path="node_modules/tns-core-modules/platform/platform.d.ts" />
/// <reference path="node_modules/tns-core-modules/text/formatted-string.d.ts" />
/// <reference path="node_modules/tns-core-modules/text/span.d.ts" />
/// <reference path="node_modules/tns-core-modules/text/text.d.ts" />
/// <reference path="node_modules/tns-core-modules/timer/timer.d.ts" />
/// <reference path="node_modules/tns-core-modules/trace/trace.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/action-bar/action-bar.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/activity-indicator/activity-indicator.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/animation/animation.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/animation/keyframe-animation.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/border/border.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/builder/builder.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/builder/component-builder.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/builder/special-properties.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/button/button.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/content-view/content-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/core/bindable.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/core/control-state-change.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/core/dependency-observable.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/core/proxy.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/core/view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/core/weak-event-listener.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/date-picker/date-picker.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/dialogs/dialogs.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/editable-text-base/editable-text-base.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/enums/enums.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/frame/frame.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/gestures/gestures.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/html-view/html-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/image-cache/image-cache.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/image/image.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/label/label.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/absolute-layout/absolute-layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/dock-layout/dock-layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/flexbox-layout/flexbox-layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/grid-layout/grid-layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/layout-base.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/stack-layout/stack-layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/layouts/wrap-layout/wrap-layout.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/list-picker/list-picker.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/list-view/list-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/page/page.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/placeholder/placeholder.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/progress/progress.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/proxy-view-container/proxy-view-container.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/repeater/repeater.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/scroll-view/scroll-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/search-bar/search-bar.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/segmented-bar/segmented-bar.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/slider/slider.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/styling/background.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/styling/css-selector.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/styling/font.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/styling/style-property.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/styling/styling.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/switch/switch.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/tab-view/tab-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/text-base/text-base-styler.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/text-base/text-base.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/text-field/text-field.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/text-view/text-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/time-picker/time-picker.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/transition/fade-transition.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/transition/slide-transition.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/transition/transition.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/utils.d.ts" />
/// <reference path="node_modules/tns-core-modules/ui/web-view/web-view.d.ts" />
/// <reference path="node_modules/tns-core-modules/utils/debug.d.ts" />
/// <reference path="node_modules/tns-core-modules/utils/lazy.d.ts" />
/// <reference path="node_modules/tns-core-modules/utils/types.d.ts" />
/// <reference path="node_modules/tns-core-modules/utils/utils.d.ts" />
/// <reference path="node_modules/tns-core-modules/xml/xml.d.ts" />
/// <reference path="node_modules/tns-core-modules/module.d.ts" />

/*
  Replaces previous single-line file  reference path to `node_modules/tns-core-modules/tns-core-modules.d.ts`
	Everything above this from `node_modules/tns-core-modules/tns-core-modules.base.d.ts` with the following exceptions:
		Removed `declarations.d.ts`
		Added `module.d.ts`
	Everything below this line is from `declarations.d.ts` but only enabled selectively to avoid `TS2300: Duplicate identifier` errors
 */


/* tslint:disable:no-unused-variable */
//
// declare class Request {
// 	constructor(input: string|Request, init?: RequestInit);
// 	method: string;
// 	url: string;
// 	headers: Headers;
// 	context: RequestContext;
// 	referrer: string;
// 	mode: RequestMode;
// 	credentials: RequestCredentials;
// 	cache: RequestCache;
// }
//
// interface RequestInit {
// 	method?: string;
// 	headers?: HeaderInit|{ [index: string]: string };
// 	body?: BodyInit;
// 	mode?: RequestMode;
// 	credentials?: RequestCredentials;
// 	cache?: RequestCache;
// }
//
// declare enum RequestContext {
// 	"audio", "beacon", "cspreport", "download", "embed", "eventsource", "favicon", "fetch",
// 	"font", "form", "frame", "hyperlink", "iframe", "image", "imageset", "import",
// 	"internal", "location", "manifest", "object", "ping", "plugin", "prefetch", "script",
// 	"serviceworker", "sharedworker", "subresource", "style", "track", "video", "worker",
// 	"xmlhttprequest", "xslt"
// }
//
// declare enum RequestMode { "same-origin", "no-cors", "cors" }
// declare enum RequestCredentials { "omit", "same-origin", "include" }
// declare enum RequestCache { "default", "no-store", "reload", "no-cache", "force-cache", "only-if-cached" }
//
// declare class Headers {
// 	append(name: string, value: string): void;
// 	delete(name: string): void;
// 	get(name: string): string;
// 	getAll(name: string): Array<string>;
// 	has(name: string): boolean;
// 	set(name: string, value: string): void;
// }
//
// declare class Body {
// 	bodyUsed: boolean;
// 	/*
// 	 arrayBuffer(): Promise<ArrayBuffer>;
// 	 blob(): Promise<Blob>;
// 	 */
// 	formData(): Promise<FormData>;
// 	json(): Promise<any>;
// 	text(): Promise<string>;
// }
//
// declare class Response extends Body {
// 	constructor(body?: BodyInit, init?: ResponseInit);
// 	error(): Response;
// 	redirect(url: string, status: number): Response;
// 	type: ResponseType;
// 	url: string;
// 	status: number;
// 	ok: boolean;
// 	statusText: string;
// 	headers: Headers;
// 	clone(): Response;
// }
//
// declare enum ResponseType { "basic", "cors", "default", "error", "opaque" }
//
// declare class ResponseInit {
// 	status: number;
// 	statusText: string;
// 	headers: HeaderInit;
// }
//
// declare type HeaderInit = Headers|Array<string>;
// declare type BodyInit = Blob|FormData|string;
// declare type RequestInfo = Request|string;
//
// declare function fetch(url: string, init?: RequestInit): Promise<Response>;
//
// interface XMLHttpRequest {
// 	send(data?: FormData): void;
// }
//
// interface Console {
// 	time(reportName: string): void;
// 	timeEnd(reportName: string): void;
// 	assert(test: boolean, message: string, ...formatParams: any[]): void;
// 	info(message: any, ...formatParams: any[]): void;
// 	warn(message: any, ...formatParams: any[]): void;
// 	error(message: any, ...formatParams: any[]): void;
// 	log(message: any, ...formatParams: any[]): void;
// 	trace(): void;
// 	dump(obj: any): void;
// 	createDump(obj: any): string;
// 	dir(obj: any): void;
// }
//
// declare var console: Console;
declare var require: NativeScriptRequire
//
// // Global functions
declare function Deprecated(target: Object, key?: string | symbol, value?: any): void
// declare function Experimental(target: Object, key?: string | symbol, value?: any): void;
//
// /**
//  * Decorates class that extends native Java class
//  * @param nativeClassName The name of the newly generated class. Must be unique in the application.
//  */
// declare function JavaProxy(nativeClassName: string): ClassDecorator;
//
// /**
//  * Important: Not applicable to Objective-C classes (iOS platform)
//  * Decorates class that extends native Java class
//  * @param interfaces An array of fully-classified Java interface names that the class must implement.
//  */
// declare function Interfaces(interfaces: any[]): ClassDecorator;
//
// declare function Log(data: any): void;
// declare function log(data: any): void;
// declare function fail(data: any): void;
//
// declare var __dirname: string;
// declare var __filename: string;
//
// /**
//  * Calls a function after a specified delay.
//  * @param callback The function to be called.
//  * @param milliseconds The time to wait before the function is called. Defaults to 0.
//  */
// declare function setTimeout(callback: Function, milliseconds?: number): number;
//
// /**
//  * Clears the delay set by a call to the setTimeout function.
//  * @param id The identifier returned by the previously called setTimeout() method.
//  */
// declare function clearTimeout(id: number): void;
//
// /**
//  * Calls a function repeatedly with a delay between each call.
//  * @param callback The function to be called.
//  * @param milliseconds The delay between each function call.
//  */
// declare function setInterval(callback: Function, milliseconds?: number): number;
//
// /**
//  * Clears repeated function which was set up by calling setInterval().
//  * @param id The identifier returned by the setInterval() method.
//  */
// declare function clearInterval(id: number): void;
//
// declare function zonedCallback(callback: Function): Function;
//
// declare class WeakRef<T> {
// 	constructor(obj: T);
// 	get(): T;
// 	clear(): void;
// }
//
declare var module: NativeScriptModule
// // Same as module.exports
// declare var exports: any;
//
// interface Array<T> {
// 	filter<U extends T>(pred: (a: T) => a is U): U[];
// }
```
